CHALLENGE INSTRUCTIONS:
- Each exercise contains one or more bugs
- Some bugs are subtle and require careful analysis
- Read error messages carefully - they're your best friend!
- Some exercises have multiple bugs - fix them all!
- Try to understand WHY the bug occurred, not just HOW to fix it

Good luck!

**CHALLENGE 1: The Mysterious Growing List**

In [3]:
def add_student(name, class_list=[]):
    """Add a student to a class list"""
    class_list.append(name)
    return class_list


print("\n--- TEST CASES ---")
print("Test 1: Class A:", add_student("Alice"))
print("Test 2: Class B:", add_student("Bob"))
print("Test 3: Class C:", add_student("Charlie"))


--- TEST CASES ---
Test 1: Class A: ['Alice']
Test 2: Class B: ['Alice', 'Bob']
Test 3: Class C: ['Alice', 'Bob', 'Charlie']


**EXPECTED OUTPUT**

"Class A: ['Alice']"

"Class B: ['Bob']

"Class C: ['Charlie']"

**CHALLENGE 2: The Mysterious Growing List**

In [30]:
total = 0

def add_to_total(value):
    # This creates a LOCAL variable named 'total'
    # instead of modifying the global one
    try:
        total= 0
        total = total + value
        return total
    except UnboundLocalError as e:
        return f"ERROR: {e}"

In [31]:
total_v2 = 0
print(f"Initial total: {total_v2}") 
total_v2 = add_to_total(10)
print(f"After adding 10: {total_v2}")
total_v2 = add_to_total(30)
print(f"After adding 20: {total_v2}")
print(f"Gloal Total: {total_v2}")

Initial total: 0
After adding 10: 10
After adding 20: 30
Gloal Total: 30


**EXPECTED OUTPUT**


"Initial total: 0"

"After adding 10: 10"

"After adding 20: 30"

"Global total: 30"

**CHALLENGE 3: Analyze Test Scores**

In [11]:
def analyze_test_scores(scores):
    total = 0
    for score in scores:
        total += score

    average = total / len(scores)
    highest = scores[0]
    lowest = scores[0]
    passing_count = 0

    for i in range(len(scores)):
        if scores[i] > highest:
            highest = scores[i]
        if scores[i] < lowest:
            lowest = scores[i]
        if scores[i] >= 60:
            passing_count += 1

    return {
        'average': average,
        'highest': highest,
        'lowest': lowest,
        'passing': passing_count,
        'failing': len(scores) - passing_count
    }

In [19]:
print("\n--- TEST CASES ---")
test_scores = [85, 92, 78, 90, 65, 88, 76, 95, 82, 70]
print(f"Input: {test_scores}")

stats = analyze_test_scores(test_scores)
print(f"Result: {stats}")


--- TEST CASES ---
Input: [85, 92, 78, 90, 65, 88, 76, 95, 82, 70]
Result: {'average': 82.1, 'highest': 95, 'lowest': 65, 'passing': 10, 'failing': 0}


{
    'average': 82.1,

    'highest': 95,

    'lowest': 65,

    'passing': 10,
    
    'failing': 0
}

**CHALLENGE 4: Dictionary Iteration Modification**

In [32]:
def remove_negative_prices(inventory):
    for item, price in list (inventory.items()):
        if price < 0:
            del inventory[item]
    return inventory


print("\n--- TEST CASES ---")
products = {
    'apple': 1.50,
    'banana': -0.50,
    'orange': 2.00,
    'grape': -1.00,
    'mango': 3.00
}



cleaned = remove_negative_prices(products.copy())
print(f"Result: {cleaned}")


--- TEST CASES ---
Result: {'apple': 1.5, 'orange': 2.0, 'mango': 3.0}


**EXPECTED OUTPUT**


Result: {'apple': 1.50, 'orange': 2.00, 'mango': 3.00}